crates/codspeed/src/fifo.rs

```diff
@@ -0,0 +1,201 @@
+use nix::libc::O_NONBLOCK;
+use nix::sys::stat;
+use nix::unistd::{self, unlink};
+use serde::{Deserialize, Serialize};
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Write};
+use std::os::unix::fs::OpenOptionsExt;
+use std::path::PathBuf;
+
+pub const RUNNER_CTL_FIFO: &str = "/tmp/runner.ctl.fifo";
+pub const RUNNER_ACK_FIFO: &str = "/tmp/runner.ack.fifo";
+
+pub struct PerfGuard {
+    ctl_fifo: FifoIpc,
+    ack_fifo: FifoIpc,
+}
+
+impl PerfGuard {
+    pub fn new(ctl_fifo: &str, ack_fifo: &str) -> Option<Self> {
+        let mut instance = Self {
+            ctl_fifo: FifoIpc::connect(ctl_fifo)?.with_writer().ok()?,
+            ack_fifo: FifoIpc::connect(ack_fifo)?.with_reader().ok()?,
+        };
+        instance.send_cmd(Command::StartBenchmark)?;
+        Some(instance)
+    }
+
+    fn send_cmd(&mut self, cmd: Command) -> Option<()> {
+        self.ctl_fifo.send_cmd(cmd)?;
+        self.ack_fifo.wait_for_ack()?;
+
+        Some(())
+    }
+}
+
+impl Drop for PerfGuard {
+    fn drop(&mut self) {
+        self.send_cmd(Command::StopBenchmark);
+    }
+}
+
+pub struct FifoIpc {
+    path: PathBuf,
+    reader: Option<File>,
+    writer: Option<File>,
+}
+
+impl FifoIpc {
+    pub fn connect<P: Into<PathBuf>>(path: P) -> Option<Self> {
+        let path = path.into();
+
+        if !path.exists() {
+            return None;
+        }
+
+        Some(Self {
+            path,
+            reader: None,
+            writer: None,
+        })
+    }
+
+    pub fn create(path: &str) -> Option<Self> {
```

Comment:
Very nitpicky: we should put this function as the first one in the impl block because it's most likely the first one the user will call